### ISSUE TITLE
Adopt 1.18 generic feature.

### ISSUE THREAD

**Issue by wdhongtw** [porfirion/trie#1]
Hi @porfirion 

I would like to submit a PR to make this library generic-able. By turning most part of the library into generic type/function and make a alias for backward-compatibility.

For example:

```golang
// Turn Trie into TrieGen with type parameter
type TrieGen[T any] struct {
	Prefix   []byte
	Value    *T // Here we use *T, not  T, for optional value
	Children *[256]*TrieGen[T]
}
// Other methods on TrieGen are modified too, with input/output type T, not *T

// Provide existing interface Trie as a type alias, ValueType will be removed directly
type Trie = TrieGen[interface{}] 
```

This interface provide easier, less error-prone interface for most types, since user don't need to 1. type assert and 2. dereference pointer.

```golang
func SomeJob() {
	var routes *TrieGen[string] = BuildFromMap[string](map[string]string{
		"/user/info": "some-user-info",
	})
	content, _, ok := routes.SearchPrefixInString("/user")
	if ok {
		fmt.Println(content)
	}
}
```

## Problem

However, the `Trie.Value` now has incompatible type chage, from `interface{}` to `*interface{}`.

Since `Value` is a public field for `Trie`, there may be some existing user who interact with the trie content,
not using public methods from the trie. For these users, the proposed change can broken their code when they upgrade this library.

Do you think this is a good change for this library? If so, I would like to submit a PR for this change. :D
(Actually I already complete the required modification and fix corresponding test cases in my local repo.)

Hope for your kind response.
And thanks for this great library, I learned a lot from it. :D

---

If we do not care about compatibility, we can just provide:

```golang
type Trie[T any] struct {
	Prefix   []byte
	Value    *T
	Children *[256]*Trie[T]
}
```

## Related Information

There is already a fork doing similar change, but the public method of that fork return values in `*T`, which I don't think a good design decision.

See: <https://github.com/MrMelon54/trie>

**Comment by porfirion** [comment]
Hi @wdhongtw!
I'm very glad that you find this library useful and it's a little surprise for me, since I didn't check it's usage for a long time :)

Yeah, I thinks time to make it generic has come. But since there are some projects depending on this library, I'd like to keep backward compatibility. Also - just making existing tree an alias for generic structure will break projects that are not using go1.18+ (like https://github.com/kotet/akizuki-bot).

Also about your concern on using `*T` as returned type of `Get*` methods: 
- on the one hand - it can be very annoying to get pointers to primitive values (or values that are already a pointers like slices and maps).
- but on the other hand - copying large structures (as I do in my projects) as values can potentially cause lots of allocations on `Get`

I think there can be a compromise: there can be two variants of the same method, one for values and one for pointers (`Get` and `GetPtr`) - it will allow users to choose preferred behaviour by themselves and wouldn't limit them.

I need some time to thinks on it and test different approaches. Will come back with decision soon :)

**Comment by paulo-raca** [comment]
Aha!
I just found this library and was starting to go through the generics changes myself, but I'm glad I found this. Thank you, @wdhongtw!

> Yeah, I thinks time to make it generic has come. But since there are some projects depending on this library, I'd like to keep backward compatibility. Also - just making existing tree an alias for generic structure will break projects that are not using go1.18+ (like https://github.com/kotet/akizuki-bot).

I think it is pretty clear that adopting generics needs to breaking, but this is probably ok: `go.sum` should take care of freezing the version in existing projects. 

Since it will breaking compatibility anyway:
- maybe the Trie attributes should all be private? :thinking: 
- I would just force generics and get rid of the `type Trie = TrieGen[interface{}]` :grimacing:

> Also about your concern on using *T as returned type of Get* methods:

I personally prefer returning `T` instead of `*T`. I'm probably going to use `Trie[*LargeStruct]` anyway whenever I'm concerned about copying structures.

However I do like the idea of having both `Get` and `GetPtr` :+1:
